src/main/java/de/tum/in/www1/artemis/domain/scoring/ScoringStrategyShortAnswerProportionalWithPenalty.java

```diff
@@ -0,0 +1,82 @@
+            double correctSolutions = 0;
+            double incorrectSolutions = 0;
+            Set<ShortAnswerSolution> solutionsAlreadyUsed = new HashSet<>();
+
+            // iterate through each spot and compare its correct solutions with the submitted texts
+            for (ShortAnswerSpot spot : saQuestion.getSpots()) {
+                Set<ShortAnswerSolution> solutionsForSpot = saQuestion.getCorrectSolutionForSpot(spot);
+                for(ShortAnswerSubmittedText text : saAnswer.getSubmittedTexts()){
+                    foundCorrectSolution = false;
+                    if(text.getSpot().equals(spot)){
+                        for(ShortAnswerSolution solution : solutionsForSpot){
+                            /*submittedText answers can be the same (more than once)
+                            when spots do not exist more than once*/
+                            if(text.isSubmittedTextCorrect(text.getText(), solution.getText())
+                                && (!saAnswer.submittedTextMoreThanOnceInSubmittedAnswer(text) || !saQuestion.spotMoreThanOnceInMapping(spot))
+                                ){
+                                text.setIsCorrect(true);
+                                correctSolutions++;
+                                foundCorrectSolution = true;
+                                break;
+                            }
+                            /*submittedText answers cannot be the same (more than once) */
+                            if(text.isSubmittedTextCorrect(text.getText(), solution.getText())
+                                && saAnswer.submittedTextMoreThanOnceInSubmittedAnswer(text)
+                                && saQuestion.spotMoreThanOnceInMapping(spot)
+                                && saQuestion.solutionMoreThanOnceInMapping(solution)
+                                && !solutionsAlreadyUsed.contains(solution)
+                                ){
+                                solutionsAlreadyUsed.add(solution);
+                                text.setIsCorrect(true);
+                                correctSolutions++;
+                                foundCorrectSolution = true;
+                                break;
+                            }
+                        }
+                        if(!foundCorrectSolution){
+                            incorrectSolutions++;
+                        }
+                    }
+                }
+            }
+
+            // calculate the fraction of the total score the user should get
+            // every correct mapping increases fraction by 1/mappedDropLocations,
+            // every incorrect mapping decreases fraction by 1/mappedDropLocations
+            double fraction = ((correctSolutions / totalSolutions) - (incorrectSolutions / totalSolutions));
+
+            // end result is maxScore * fraction, but at least 0
+            return Math.max(0, question.getScore() * fraction);
+        } else {
```

Comment:
You don't need an `else` branch if you `return` in the if

Reply:
Done, I will adapt it for MC and DnD as well. :)